src/Testing/ZopeTestCase/functional.py

```diff
@@ -80,13 +81,11 @@ def publish(self, path, basic=None, env=None, extra=None,
         env['SERVER_PROTOCOL'] = 'HTTP/1.1'
         env['REQUEST_METHOD'] = request_method
 
-        p = path.split('?')
-        if len(p) == 1:
-            env['PATH_INFO'] = p[0]
-        elif len(p) == 2:
-            [env['PATH_INFO'], env['QUERY_STRING']] = p
-        else:
-            raise TypeError('')
+        query = ''
+        if '?' in path:
+            path, query = path.split("?", 1)
+        env['PATH_INFO'] = unquote_to_bytes(path).decode('latin-1')
```

Comment:
In my opinion, it is conceptually wrong to "unquote" here: the path typically has a substructure; unquoting the complete path (rather than the individual components) can lose information necessary to correctly identify the components (e.g. in the case of quoted "/" or ";" to indicate their use as literal part of a component rather than as separator).
However, the handling is wrong in other `Zope` parts; thus, it is likely acceptable to be wrong here, too.

Reply:
Thanks for the feedback @d-maurer 

I agree with this point. During testing, I also noticed that it was not possible in Zope (not only in the testing utilities but also with real requests) to use a path such as `a/b/c%2Fd` and have it traversed as `['a', 'b', 'c/d'] - but this is not only Zope doing it wrong here, this is also in the WSGI spec.

Apparently, ASGI has [`raw_path`](https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope) for that use case. Some WSGI servers put the raw URI or path in environment ( for example https://github.com/Pylons/waitress/issues/346 ). To be correct, Zope could try to use these raw URI instead of the standard `PATH_INFO`, but in practice even if the current way is not correct, I guess it's not a problem affecting many people.

As you said, it's not only this part, with these changes we are making the testing utilities have a consistent behavior with the real requests, you have approved the changes, so I am going to merge. Thanks again !

I would also appreciate feedback on https://github.com/zopefoundation/Zope/pull/1060 . Does it make sense to backport these kinds of fixes on Zope4 ? if it's in Zope4 I would use it, otherwise I would use a monkey patch in my application.